### PR TITLE
Proximity

### PR DESCRIPTION
**Pull request by j042**
# Overview
(WIP) working prototype of finding closest parametric coordinates using so called __j<sup>2</sup>(f<sub>0</sub>)__ algorithm.
Will explain more before merge.

### CODE REVIEW

**Comment by j042** [comment]
@jzwar

File: src/splinelib_ext/arrutils.hpp [hunk 1]
```diff
@@ -0,0 +1,389 @@
...
+      maybemax = std::abs(A[j][i]);
+      if (maybemax > maxval) {
+        maxval = maybemax;
+        maxrow = j;
+      }
+    }
+    // swap if needed. max entry row goes to i-th row.
+    if (maxrow != i) {
+      // swap matrix entries
+      // TODO: since it is nested, we can just do once?
```
Comment (j042) [inline src/splinelib_ext/arrutils.hpp:250]: yes

File: src/splinelib_ext/arrutils.hpp [hunk 2]
```diff
@@ -0,0 +1,389 @@
...
+    // ignore clipped entries
+    if (skipmask[i] != 0) continue;
+
+    sum = 0.;
+    for (int j{i+1}; j < para_dim; j++) {
+      sum += A[i][j] * x[j];
+    }
+    x[i] = (b[i] - sum) / A[i][i];
+    
+    std::cout << "x" << i <<": " << x[i] << " ";
```
Comment (j042) [inline src/splinelib_ext/arrutils.hpp:290]: rm

File: src/splinelib_ext/nurbspl.hpp [hunk 3]
```diff
@@ -139,6 +217,378 @@ class NurbsExt : public splinelib::sources::splines::Nurbs<para_dim, dim> {
...
+
+        // fill symmetric part
+        if (i != j) lhs[j][i] = lhs[i][j];
+      }
+    }
+  }
+
+  /***************************************************************************/
+  /********** [end] Helper functions for `FindParametricCoordinate` **********/
+  /***************************************************************************/
```
Comment (jzwar) [inline src/splinelib_ext/nurbspl.hpp:467]: What the `C` is that??

File: src/splinelib_ext/nurbspl.hpp [hunk 4]
```diff
@@ -139,6 +217,378 @@ class NurbsExt : public splinelib::sources::splines::Nurbs<para_dim, dim> {
...
+      _build_d2jdu2(current_guess, dist, eyeders, d2jdu2); /* lhs */
+
+      /* debug prints */
+      std::cout << "**************** Newton iteration: " << i << " **********"; 
+      std::cout << "\n distance (goal - spline(para_c): ";
+      for (int i{0}; i < para_dim; i++) {
+        std::cout << dist[i] << " ";
+      }
+      std::cout << "\n";
+      std::cout << "System Matrix: A | b  (d2jdu2 | djdu)\n";
```
Comment (jzwar) [inline src/splinelib_ext/nurbspl.hpp:526]: Is user output necessary

File: src/splinelib_ext/nurbspl.hpp [hunk 5]
```diff
@@ -139,6 +217,378 @@ class NurbsExt : public splinelib::sources::splines::Nurbs<para_dim, dim> {
...
+
+      // Customer satisfaction survey
+      // ============================
+      // 1. Converged?
+      //   -> should be zero, if query point is on the spline.
+      // 2. All clipped?
+      //   -> should be the case if query point lies outside the spline.
+      curnorm = norm2(djdu);
+      std::cout  << "\n norm2(djdu): " << curnorm << "\n";
+      if (std::abs(prevnorm - curnorm) < tolerance
```
Comment (jzwar) [inline src/splinelib_ext/nurbspl.hpp:584]: Will never be fulfilled on first iteration. Although the first iteration can be sufficient. If

**Review by jzwar — CHANGES_REQUESTED** [review]
Some additional points
1. There seems to be no use of the newly provided functions in spinpy's python interface.
2. Templatize all functions please
3. Revisit the c++ naming conventions

File: src/helpers.hpp [hunk 6]
```diff
@@ -0,0 +1,119 @@
+#pragma once
+
+#include <cmath>
+#include <thread>
+#include <utility>
+
+
+/* thread for all */
+template <typename Func, typename IndexT>
+void nthread_execution(
```
[the remark this answers is not in the document]
Comment (j042) [reply src/helpers.hpp:10]: `MultiThreadExecutionButIfGivenNThreadIsEqualToOneItWillNotCreateThreads`

File: src/helpers.hpp [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+    return;
+  }
+
+  // get chunk size and prepare threads
+  const IndexT chunk_size = std::ceil(total / nthread);
+  std::vector<std::thread> tpool;
+  tpool.reserve(nthread);
+
+  for (int i{0}; i < (nthread - 1); i++) {
+    tpool.emplace_back(std::thread{f, i * chunk_size, (i + 1)*chunk_size});
```
[the remark this answers is not in the document]
Comment (j042) [reply src/helpers.hpp:27]: taf!

File: src/helpers.hpp [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+  }
+  {
+    // last one
+    tpool.emplace_back(
+        std::thread{f, (nthread - 2) * chunk_size, total}
```
[the remark this answers is not in the document]
Comment (j042) [reply src/helpers.hpp:32]: good catch, thanks

File: src/helpers.hpp [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+  }
+}
+
+/* raster point generator */
+template <int dim, typename FillType>
+void raster_points(
+    
+) {
+}
+
```
[the remark this answers is not in the document]
Comment (j042) [reply src/helpers.hpp:47]: k

**Comment by j042** [comment]
thanks for the comments, I will work on those! closing this one for now.

**Pull request closed without merging**